### DISCUSSION TITLE
Explain features: Deep & Heuristic mode

### DISCUSSION THREAD

**Discussion by NtWriteCode** [horsicq/Detect-It-Easy#232] [General]
Are there any documentations about these features? I can't find any info on how they work. I'm just curious on the high level if signatures work differently or DIE scans somehow differently or what else?

Thank you! :)

**Comment by horsicq** [comment]
Some signatures use "isDeepScan()" and "isHeuristicScan()" functions. It works if we use the checkboxes or flags

**Comment by horsicq** [comment]
Example: https://github.com/horsicq/Detect-It-Easy/blob/6cba6b3f3440c7f916386e33ad700f70de4b2bb1/db/PE/ASM_Guard.2.sg#L11
Example: https://github.com/horsicq/Detect-It-Easy/blob/6cba6b3f3440c7f916386e33ad700f70de4b2bb1/db/PE/__GenericHeuristicAnalysis_By_DosX.7.sg#L74

**Comment by Kaens** [comment]
Here's the way I've come to be using these...

There are some files whose signature you just can't be sure about.
You can make the detection more involved than the all-too-common, "if starts with MZ then an .exe file".
If the user chooses to DeepScan, you can spend some extra CPU time.
 For example, you've a format that starts with "MZ"; you could follow into the format's own bytecode and looking at each command to sanity-check whether the loop bytecode doesn't overshoot the file beginning, or the note fine-tune is within -15～+15, or whether the next block starts right after the previous block's "end of block" bytecode is encountered. Then you can tell with confidence it's a file of a specific format, or it can be used as such (almost the same thing amirite).

And if the user chooses to HeuristicScan, you can see they want you to maybe be a bit more lenient in detection, in case the format has a good amount of outliers that don't strictly follow the expected format descriptions.
 Like, in the same "MZ" format, you could already tell the user that this might be an .exe file; so your detector would have something like this (bad practice but serves as a good example):
```js
if (File.isHeuristicScan() && File.compare("'MZ'"))
  _setResult("~executable", "MSDOS/Windows executable (.EXE)", "", "");
```
(where the starting tilde means you're not sure about this detection; DIE parses that)

In addition, isAggressiveScan() could allow you to have a clear conscience for even more, like you could unpack a custom LZX in the file and extract some details, all without the user having to launch the software. It would take much more CPU time, but is it worth it? The user says so!

**Reply by NtWriteCode** [reply]
Amazing extra addition/description, thank you!

**Comment by DosX-dev** [comment]
> I can't find any info on how they work

Source of DiE PE heuristic analyzer -> [click](https://github.com/horsicq/Detect-It-Easy/blob/master/db/PE/__GenericHeuristicAnalysis_By_DosX.7.sg)
